### PR TITLE
patch-headers: Add fixup hook

### PR DESCRIPTION
**Pull request by wkennington**
This patch is the first step in removing some of the propagatedBuildInputs hacks needed for some applications which depend on headers but don't list them in their package config, or are not detected through the package config interface.

This patch adds a hook which resolves all includes within c / c++ header files to their absolute paths and substitutes them into the original headers. This should mimic the behavior of cpp -I / -isystem in the behavior of resolution. It also respects the distinction between system includes and system + cwd includes (#include <> vs #include "").

Eg. If we are building libvorbis and we have $out/include/codec.h which contains "#include <ogg/ogg.h>" this would resolve that to ${libogg}/include/ogg/ogg.h and replace the include with "#include \"/nix/store/xxxxxxxxxxxxxxxxxx-libogg-1.3.2/include/ogg/ogg.h\"".

### CODE REVIEW

**Comment by wkennington** [comment]
@edolstra @vcunat @lethalman @domenkozar

**Comment by vcunat** [comment]
I don't see why ``propagatedBuildInputs`` is considered a hack, and this PR seems even more hackish to me (without long thinking yet). Besides, we typically need corresponding libraries as well, which is easiest with ``propagatedBuildInputs`` again AFAIK. (It's probably too soon to discuss into which branch this PR might be merged.)

**Comment by wkennington** [comment]
@vcunat I'm working on the fix for .la and .pc files now. The problem with propagatedBuildInputs is that it becomes hard to follow exactly what dependencies a package is using when you have lots of propagatedBuildInputs from dependencies.

**Comment by vcunat** [comment]
Hmm, making the propagated dependencies more precise, perhaps. But it's another departure from the usual behavior, so we might hit some new bugs in packages (their configure+build systems).

I expect we may have to be careful about what these things do during stdenv bootstrapping. Off the top of my head, I don't see any other case where we might want to "change dependencies" in a similar way (and ``replaceDependency`` should be fine with this).

**Comment by wkennington** [comment]
The only problem I can see now is with the bootstrap tools themselves. This
seems to work without problem during the full rebuild it causes.

**Comment by edolstra** [comment]
I'm basically +1 on the idea, though I'm not sure whether we should be doing this in bash. It's a terrible (and slow) language if you're doing anything complicated, like string manipulation on lots of files. If you're going to generate a call to `awk`, you may as well do the whole thing in Perl :-)

@vcunat `propagatedBuildInputs` is kind of undesirable because it leads to implicit build-time dependencies, which might lead to unexpected build failures or missing features, e.g. if you have something like
```
buildInputs = [ foo ] ++ optional enableBar bar;
```
If this package has an implicit dependency on `xyzzy`, which is propagated by `bar`, then it will suddenly fail if `enableBar` is set to false.

It's kind of like `RPATH` vs `LD_LIBRARY_PATH` - the first is preferable in Nix because it doesn't require the user of a package to set up the environment.

**Comment by edolstra** [comment]
Though of course Perl is not part of `stdenv`...

**Comment by wkennington** [comment]
I mean, bash is slow, especially since I can't nicely cache header lookups
or anything terribly fancy. However, it wasn't that slow when I rebuilt the
stdenv and some core packages. My guess is that it would only be noticeably
bad for gcc and boost. Worst case we could write a binary for this one but
I dont think this routine is slow enough to justify that yet.

**Comment by wkennington** [comment]
This still needs support for system headers, having spaces within the macro definition, as well as supporting include_next. The first two things would be easy to do with the current script, but supporting include_next might require some extra thinking on the matter. It seems like include_next is mostly used in system files and really isn't a big deal for a majority of headers.

**Comment by wkennington** [comment]
This now supports patching the libc and gcc headers with a few exceptions needed so that gcc and glibc do not depend on bootstrap-utils. Everything looks solid except for the derivations like bootstrap-tools which take include paths outside the store.

**Comment by wkennington** [comment]
I think at this point someone should spin up a hydra jobset with a minimal set of packages in order to validate the behavior is as expected. I'd like to cut this over to staging as soon as possible.

**Comment by lethalman** [comment]
Before merging this I'd like to see a better understanding of the consequences of this. I mean, it's ok for me, but are we forgetting any use case? It's easy from our view point, but are we cutting any possible other users use case in the future? I'd invite to think more about any possible problems that we're overseeing.

Are there any use cases where headers shall not be absolute paths?

**Comment by jagajaga** [comment]
ping all

**Comment by wkennington** [comment]
This can be closed
